### PR TITLE
Orderbook foundation

### PR DESCRIPTION
**Pull request by bonomat**
This little project makes use of

- actix: so that we can use websockets
- diesel: we wanted to use diesel over sqlx due to performance issue and other annoyances
- postgres: we want to use postgres for our services to be able to use db as a service
- docker: we need to run a postgres db
- testcointainers: we also need to run a db for our tests, hence, we use our beloved testcontainers for this

addresses: https://github.com/get10101/10101/issues/156

### CODE REVIEW

**Comment by holzeis** [comment]
@bonomat Are now going for a separate order book service after all? or is this intended to get merged with the coordinator?

**Comment by bonomat** [comment]
> @bonomat Are now going for a separate order book service after all? or is this intended to get merged with the coordinator?

I would merge it with the coordinator once we are happy with it. I think it's so nicely decopelled that we can work in it in a separate binary :).

**Comment by bonomat** [comment]
@da-kami / @Restioson / @luckysori :if you want you can have already a look. 

I don't expect an indepth review. Afaik it's the first time we are using actix and diesel, so there might be a few things which we can do better. Also error handling has not been taken care yet. 

I plan to add websockets tomorrow then it's ready for full review.

**Comment by Restioson** [comment]
Have not used actix-web in a long time but I have used it for a web socket app in the past. I think nowadays people also like to use warp and axum which also handle websockets. Back then I switched off of actix-web because of the unsoundness in actix's codebase but I believe the new maintainers have addressed that now.

File: orderbook/src/bin/orderbook.rs
Comment (Restioson) [inline orderbook/src/bin/orderbook.rs:27]: "mysecretpassword" should contain a special character, uppercase, and number - try "My$3cretpassword" :wink:

File: orderbook/src/bin/orderbook.rs
Comment (Restioson) [inline orderbook/src/bin/orderbook.rs:50]: I wonder if there is any tracing middleware for actix :thinking: if we want to go that route again, we should consider ahead of time how much detail we want to include in the traces so we avoid explosive disk usage

File: orderbook/src/routes/mod.rs
Comment (Restioson) [inline orderbook/src/routes/mod.rs:76]: This is something we need to be very careful of in general; in the past I've gone as far as only allowing access to the DB through a `database` model which wraps every DB call in `async` first. Otherwise maybe we can lint or something. It is just very easy to forget one DB call on the reactor thread and then get weird performance issues elsewhere (as seen in ItchySats)

File: orderbook/src/routes/session.rs
Comment (Restioson) [inline orderbook/src/routes/session.rs:75]: Does actix still not allow Handlers and lifecycle methods to be async blocks? :grimacing:

**Review by Restioson — COMMENTED** [review]
Looks good so far! :rocket:

File: orderbook/src/bin/orderbook.rs
Comment (bonomat) [reply orderbook/src/bin/orderbook.rs:27]: 😅

File: orderbook/src/bin/orderbook.rs
Comment (bonomat) [reply orderbook/src/bin/orderbook.rs:50]: Looks like there is a lib for it. 
I agree, better look into this sooner than later :)

File: orderbook/src/routes/session.rs
Comment (bonomat) [reply orderbook/src/routes/session.rs:75]: Does not look like it.

**Comment by 1010Tom** [comment]
@Restioson : I've reimplented it using Axum and it feels much faster :)

**Comment by 1010Tom** [comment]
@luckysori / @Restioson  / @da-kami : if we are happy with this approach then I would merge it into `coordinator`.

**Review by da-kami — APPROVED** [review]
Looks good, happy to roll with `axum` - don't have too much experience with different web frameworks, but it looks good and is developed by the tokio team :)

**Comment by da-kami** [comment]
Note: On the `coordintator` side we are using rocket at the moment. I don't think it is efficient to use two different web frameworks so I would adapt that if we decide to use `axum` - CC @holzeis

**Comment by bonomat** [comment]
I'll merge this for now, then go ahead and change the coordinator to use `axum` and then migrate the orderbook to the coordinator :)

**Comment by bonomat** [comment]
@1010Tom : I'll take over

File: orderbook/src/routes/session.rs
Comment (Restioson) [reply orderbook/src/routes/session.rs]: Very sad

**Pull request merged**
